### PR TITLE
gpu objects use smart pointer

### PR DESCRIPTION
**Pull request by bluesky013**
Re: #

### Changelog

* gpu objects use smart pointer
* update bufferView binding in descriptor set after buffer resize

-------

### Continuous Integration

This pull request:

* [x] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Comment by bluesky013** [comment]
@cocos-robot run test cases

File: native/cocos/renderer/gfx-vulkan/VKCommands.h [hunk 1]
```diff
@@ -48,17 +48,10 @@ void cmdFuncCCVKCreateComputePipelineState(CCVKDevice *device, CCVKGPUPipelineSt
...
+void cmdFuncCCVKCopyTextureToBuffers(CCVKDevice *device, const CCVKGPUTexture *srcTexture, const CCVKGPUBufferView *destBuffer, const BufferTextureCopy *regions, uint32_t count, const CCVKGPUCommandBuffer *gpuCommandBuffer);
 
-void cmdFuncCCVKDestroyQueryPool(CCVKGPUDevice *device, CCVKGPUQueryPool *gpuQueryPool);
-void cmdFuncCCVKDestroyRenderPass(CCVKGPUDevice *device, CCVKGPURenderPass *gpuRenderPass);
 void cmdFuncCCVKDestroySampler(CCVKGPUDevice *device, CCVKGPUSampler *gpuSampler);
-void cmdFuncCCVKDestroyShader(CCVKGPUDevice *device, CCVKGPUShader *gpuShader);
-void cmdFuncCCVKDestroyFramebuffer(CCVKGPUDevice *device, CCVKGPUFramebuffer *gpuFramebuffer);
-void cmdFuncCCVKDestroyDescriptorSetLayout(CCVKGPUDevice *device, CCVKGPUDescriptorSetLayout *gpuDescriptorSetLayout);
-void cmdFuncCCVKDestroyPipelineLayout(CCVKGPUDevice *device, CCVKGPUPipelineLayout *gpuPipelineLayout);
-void cmdFuncCCVKDestroyPipelineState(CCVKGPUDevice *device, CCVKGPUPipelineState *gpuPipelineState);
```
Comment (star-e) [inline native/cocos/renderer/gfx-vulkan/VKCommands.h]: Why are all destroy functions removed?

File: native/cocos/renderer/gfx-vulkan/VKGPUObjects.h [hunk 2]
```diff
@@ -184,61 +201,64 @@ struct CCVKGPUBuffer {
 
     // descriptor infos
     VkBuffer vkBuffer = VK_NULL_HANDLE;
-    VkDeviceSize startOffset = 0U;
     VkDeviceSize size = 0U;
 
     VkDeviceSize instanceSize = 0U; // per-back-buffer instance
-    ccstd::vector<ThsvsAccessType> currentAccessTypes;
+    mutable ccstd::vector<ThsvsAccessType> currentAccessTypes;
```
Comment (star-e) [inline native/cocos/renderer/gfx-vulkan/VKGPUObjects.h]: mutable member usually means thread-safety. See [https://softwareengineering.stackexchange.com/questions/274548/const-means-thread-safe](url)
Consider adding mutex, or changing constness.

File: native/cocos/renderer/gfx-vulkan/VKCommands.h (hunk 1 above)
Comment (bluesky013) [reply native/cocos/renderer/gfx-vulkan/VKCommands.h]: The destruction process should only be allowed in the device object destructor or recycle bin. And I missed the sampler destroy function.

**Comment by hana-alice** [comment]
How to make sure `getObjectID(Agent<GFXBuffer>* buffer)` and `getObjectID(GFXBuffer*)` keep consistent? Or is there any  mechanism makes it doesn't matter even thery are different?

**Comment by bluesky013** [comment]
> How to make sure `getObjectID(Agent<GFXBuffer>* buffer)` and `getObjectID(GFXBuffer*)` keep consistent? Or is there any mechanism makes it doesn't matter even thery are different?

There are two gfx::Buffer instance in ```Agent<gfx::Buffer>```
```cpp
gfx::Agent<gfx::Buffer> *bufferAgent = deviceAgent.createBuffer();
gfx::Buffer* buffer = static_cast<gfx::Buffer*>(bufferAgent);
gfx::Buffer* bufferActor = bufferAgent->actor;

// 1. first instance
getObjectID(bufferAgent) == getObjectID(buffer); // objectID1

// 2. second instance
getObjectID(bufferActor); // objectID2
```
Agent shoud care which one to use.

File: native/cocos/renderer/gfx-vulkan/VKGPUObjects.h (hunk 2 above)
Comment (bluesky013) [reply native/cocos/renderer/gfx-vulkan/VKGPUObjects.h]: I have chaned constness of ```TextureView``` and ```BufferView```. 

Here are 3 members should be removed later, which still has mutable specifier.
a.  ```vector<uint32_t> clearValues``` in CCVKGPURenderpass.
b. ```vector<uint32_t> dynamicOffsetOffsets``` in CCVKGPUPipelineLayout.
c. ```vector<VkBuffer> recordedBuffers``` in CCVKGPUCommandBuffer.

**Pull request closed without merging**